### ISSUE TITLE
Cooperate with VTE to provide process completion notifications

### ISSUE THREAD

**Issue by Shnatsel** [fish-shell/fish-shell#1382]
Pantheon Terminal, part of elementary OS and Pantheon DE, once supported process completion notifications as seen in https://www.youtube.com/watch?v=SyG1-Szs0YY
The implementation was a dirty hack so it was eventually removed.

I'm looking to re-introduce the feature, cleanly this time. This requires the virtual terminal emulator and the shell to cooperate, because job completion detection in terminal without shell cooperation is unreliable, and the shell without terminal cooperation would not know to which window or tab the completed process belongs, whether the relevant window is focused (and thus show superfluous notifications), etc.

I have a basic idea on how to hack this into bash via example variables. Being an avid fish user, I'd love if fish supported that as well. However, the bash approach would not work for fish (according to built-in help page anyway), and besides I'd very much prefer it to support the feature natively.

My implementation notes can be found at https://docs.google.com/document/d/1Tgf8WpKrSpZnDRXGJhY3JHNbmsCaJpnTkApmdFC92No/edit 
I'd appreciate any comments or feedback.

**Comment by faho** [comment]
>It would probably be most convenient from the terminal's perspective if it could set an environment variable with the code for FISH to execute after completion of every user-initiated command. Ideally that code should receive the command that's just completed as an argument, but looking up history works too.

This seems like a prime usecase for an eventhandler, i.e.

```
function pantheon_notify --on-event fish_postexec
dbus-send --type=method_call --session --dest=net.launchpad.pantheon-terminal /net/launchpad/pantheon_terminal org.pantheon.terminal.ProcessFinished string:$PANTHEON_TERMINAL_ID string:"$argv[1]";
end
```

Though note that this is executed after _every_ command, not just long-running ones.

**Comment by zanchey** [comment]
The post-exec hook is definitely the right way to go, but how should the terminal communicate to the shell that it's required? I'm nervous about importing functions from the environment - see e.g. shellshock. Is there a better way?

**Comment by faho** [comment]
There's two easy ways I can think of: Either let supporting terminals add the function as a file or to config.fish or add it upstream (i.e. here) and add a commandline option or special env var (e.g. "NOTIFY=1" - nothing with arbitrary code, just a value to activate the already included function) to fish. Neither are going to win beauty contests, though.

**Comment by shnatsel-unprivileged** [comment]
I've tried entering the above function definition in interactive shell, and it worked! Thanks!

However, I could not get it to work from ~/.config/fish/config.fish file - it works if I explicitly source it from interactive shell, but not on its own. I've run strace on the file and it is indeed read by the shell, but for whatever reason the callback is not executed until I explicitly source it.

Same thing with function definition - I've put that in a function file and the function does not get executed on task completion. It is displayed correctly in fish_config web interface, but is not hooked up to fish_postexec unless I explicitly source it.

Sooo close! But for the love of all that's fishy, what am I doing wrong?!

> Neither are going to win beauty contests, though.

Adding a function as a file and hooking it up to command completion event without breaking anything else is *way more* than other shells currently offer. 

For BASH we have to do a lot of hackery on a environment variable to inject a callback without breaking anything, but we still broke something that arch ships out of the box regardless and there's nothing we can do to fix that. And [don't even get me started on ZSH](https://superuser.com/questions/735660/whats-the-zsh-equivalent-of-bashs-prompt-command).

So fish's function files win both beauty and reliability contests here. If I could get them to work, that is.

**Comment by Shnatsel** [comment]
I've tried entering the above function definition in interactive shell, and it worked! Thanks!

However, I could not get it to work from ~/.config/fish/config.fish file - it works if I explicitly source it from interactive shell, but not on its own. I've run strace on the file and it is indeed read by the shell, but for whatever reason the callback is not executed until I explicitly source it.

Same thing with function definition - I've put that in a function file and the function does not get executed on task completion. It is displayed correctly in fish_config web interface, but is not hooked up to fish_postexec unless I explicitly source it.

Sooo close! But for the love of all that's fishy, what am I doing wrong?!

> Neither are going to win beauty contests, though.

Adding a function as a file and hooking it up to command completion event without breaking anything else is *way more* than other shells currently offer.

For BASH we have to do a lot of hackery on a environment variable to inject a callback without breaking anything, but we still broke something that arch ships out of the box regardless and there's nothing we can do to fix that. And don't even get me started on [the way ZSH handles this](https://superuser.com/questions/735660/whats-the-zsh-equivalent-of-bashs-prompt-command).

So fish's function files win both beauty and reliability contests here. If I could get them to work, that is.

**Comment by ridiculousfish** [comment]
Autoloading functions cannot be event handlers until they get run once. It's a chicken and egg problem - see #845.

However it's surprising that it fails in `config.fish`. Can you check if the function is defined after? `functions myfuncname`.

I tried putting this in `config.fish`:

    function somefunc --on-event fish_postexec
       echo BEEP
    end

and it did indeed echo BEEP after every command.

**Comment by Shnatsel** [comment]
That worked for me, thanks! Looks like conditional definition of that function was the culprit. For example,

    if status --is-interactive
        function somefunc --on-event fish_postexec
            echo BEEP
        end
    end

does not work, but this works:

    function somefunc --on-event fish_postexec
        if status --is-interactive
            echo BEEP
        end
    end

One last question remains, then. How do I inject an event handler without breaking anything? Something similar to a function file would be ideal because I could easily update or patch the code therein and also cleanly remove it in case our terminal gets uninstalled. Via something like config.fish.d, perhaps?

**Comment by ridiculousfish** [comment]
Unfortunately there's no mechanism yet for autoloading event handling functions. Maybe you can think of one :)

Until then we can just put it in `share/functions/__fish_config_interactive.fish`, that's where platform-specific initialization lives.

**Comment by zanchey** [comment]
Hopefully the solutions to #1956 will help solve this problem.

**Comment by TeNNoX** [comment]
It works like that (see https://gist.github.com/TeNNoX/e027e5daabe4997093a49873411f5c0e), but the problem with not knowing if the terminal is in foreground is still there. (I solved it by using an xdotool to check the currently active window - beatiful ^^)

Any progress on how that communication could happen?

**Comment by faho** [comment]
@TeNNoX: The idea in this issue is that you wouldn't call `notify-send` - which sends a dbus message to the general notification service - but you'd send a dbus message to the terminal.

See earlier comments:

>dbus-send --type=method_call --session --dest=net.launchpad.pantheon-terminal /net/launchpad/pantheon_terminal org.pantheon.terminal.ProcessFinished string:$PANTHEON_TERMINAL_ID string:"$THECOMMANDTHATFINISHED";

(getting the correct value for that last variable is left as an exercise to the reader)

**Comment by glehmann** [comment]
This feature is awesome!
For the record, I just added
~~~
function pantheon_notify --on-event fish_postexec
    if status --is-interactive
        dbus-send --type=method_call --session --dest=org.pantheon.terminal /org/pantheon/terminal org.pantheon.terminal.ProcessFinished string:$PANTHEON_TERMINAL_ID string:"$argv[1]";
    end
end
~~~
in my `~/.config/fish/config.fish`.

Is it still ok to put that code in `share/functions/__fish_config_interactive.fish`?
Would it be ok to check the content of `COMMAND_PROMPT` — which is already set by the pantheon terminal — in order to activate that function?

**Comment by faho** [comment]
With the conf.d snippet stuff that landed in 2.3.0, this is now something that can easily be done with local customization.

I kinda don't think this needs to be in fish proper.

**Comment by Shnatsel** [comment]
The addition of vendor config is great news! I can finally get pantheon-terminal to integrate with fish out of the box. Thanks!
